src/thunderbots/software/geom/util.cpp

```diff
@@ -0,0 +1,965 @@
+    }
+    // do angle sweep for largest angle
+    std::sort(events.begin(), events.end());
+    Angle best  = Angle::zero();
+    Angle sum   = Angle::zero();
+    Angle start = events[0].first;
+    int cnt     = 0;
+    for (std::size_t i = 0; i + 1 < events.size(); ++i)
+    {
+        cnt += events[i].second;
+        assert(cnt <= 1);
+        if (cnt > 0)
+        {
+            sum += events[i + 1].first - events[i].first;
+            if (best < sum)
+            {
+                best = sum;
+                // shoot ray from point p
+                // intersect with line p1-p2
+                const Angle mid    = start + sum / 2 + offangle;
+                const Vector ray   = Vector::createFromAngle(mid) * 10.0;
+                const Vector inter = lineIntersection(src, src + ray, p1, p2);
+                bestshot           = inter;
+            }
+        }
+        else
+        {
+            sum   = Angle::zero();
+            start = events[i + 1].first;
+        }
+    }
+    return std::make_pair(bestshot, best);
+}
+
+std::vector<Vector> circleBoundaries(const Vector &centre, double radius, int num_points)
+{
+    Angle rotate_amount = Angle::full() / num_points;
+    std::vector<Vector> ans;
+    Vector bound(radius, 0.0);
+    for (int i = 0; i < num_points; i++)
+    {
+        Vector temp = centre + bound;
+        ans.push_back(temp);
+        bound = bound.rotate(rotate_amount);
+    }
+    return ans;
+}
+
+bool collinear(const Vector &a, const Vector &b, const Vector &c)
+{
+    if ((a - b).lensq() < EPS2 || (b - c).lensq() < EPS2 || (a - c).lensq() < EPS2)
+    {
+        return true;
+    }
+    return std::fabs((b - a).cross(c - a)) < EPS;
+}
+
+Vector clipPoint(const Vector &p, const Vector &bound1, const Vector &bound2)
+{
+    const double minx = std::min(bound1.x(), bound2.x());
+    const double miny = std::min(bound1.y(), bound2.y());
+    const double maxx = std::max(bound1.x(), bound2.x());
+    const double maxy = std::max(bound1.y(), bound2.y());
+    Vector ret        = p;
+    if (p.x() < minx)
+    {
+        ret.set(minx, ret.y());
+    }
+    else if (p.x() > maxx)
+    {
+        ret.set(maxx, ret.y());
+    }
+    if (p.y() < miny)
+    {
+        ret.set(ret.x(), miny);
+    }
+    else if (p.y() > maxy)
+    {
+        ret.set(ret.x(), maxy);
+    }
+    return ret;
+}
+
+Vector clipPoint(const Vector &p, const Rect &r)
+{
+    const double minx = r.swCorner().x();
+    const double miny = r.swCorner().y();
+    const double maxx = r.neCorner().x();
+    const double maxy = r.neCorner().y();
+    Vector ret        = p;
+    if (p.x() < minx)
+    {
+        ret.set(minx, ret.y());
+    }
+    else if (p.x() > maxx)
+    {
+        ret.set(maxx, ret.y());
+    }
+    if (p.y() < miny)
+    {
+        ret.set(ret.x(), miny);
+    }
+    else if (p.y() > maxy)
+    {
+        ret.set(ret.x(), maxy);
+    }
+    return ret;
+}
+
+// TODO: does this work? unit test
```

Comment:
Scary `TODO`. Might as well add some unit tests

Reply:
This is actually tested, forgot to remove todo